timemachine/fe/single_topology.py

```diff
@@ -896,8 +926,102 @@ def __init__(self, mol_a, mol_b, core, forcefield):
         # setup end states
         self.src_system = self._setup_end_state_src()
+
+        print(len(self.src_system.chiral_atom.potential.idxs))
+        print(len(self.src_system.chiral_atom.params))
+        # assert 0
         self.dst_system = self._setup_end_state_dst()
 
+        self.assert_chiral_consistency_and_validity(
+            self.src_system.chiral_atom.potential.idxs,
+            self.dst_system.chiral_atom.potential.idxs,
+            self.src_system.bond.potential.idxs,
+            self.dst_system.bond.potential.idxs,
+        )
+
+    def get_neighbors(self, atom, bond_idxs):
+        nbs = []
+        for i, j in bond_idxs:
+            if i == atom:
+                nbs.append(j)
+            elif j == atom:
+                nbs.append(i)
+        return nbs
+
+    def check_chiral_validity(self, src_chiral_centers_in_mol_c, dst_canon_regular_idxs, src_bond_idxs):
+        for c in src_chiral_centers_in_mol_c:
+            nbs = self.get_neighbors(c, src_bond_idxs)
+            if len(nbs) == 4:
+                i, j, k, l = nbs
+                # (ytz): the ordering of i,j,k,l is random if we're reading directly from the mol graph,
+                # which can be inconsistent with the ordering used in the chiral volume definition.
+                # However, since we're guaranteed to never have inversions between end-states,
+                # we can just quickly check both-ways to see if the desired chiral volume is present.
+                flag_0 = (
+                    canonicalize_chiral_atom_idxs((c, i, j, k)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, j, i, k)) in dst_canon_regular_idxs
+                )
+                flag_1 = (
+                    canonicalize_chiral_atom_idxs((c, i, j, l)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, j, i, l)) in dst_canon_regular_idxs
+                )
+                flag_2 = (
+                    canonicalize_chiral_atom_idxs((c, i, k, l)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, k, i, l)) in dst_canon_regular_idxs
+                )
+                flag_3 = (
+                    canonicalize_chiral_atom_idxs((c, j, k, l)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, k, j, l)) in dst_canon_regular_idxs
+                )
+                # print("Processing chiral center", c, "with neighbors", i,j,k,l)
+
+                if not flag_0 and not flag_1 and not flag_2 and not flag_3:
+                    raise ChiralConversionError()
+
+            if len(nbs) == 3:
+                i, j, k = nbs
+                flag_0 = (
+                    canonicalize_chiral_atom_idxs((c, i, j, k)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, j, i, k)) in dst_canon_regular_idxs
+                )
+                if not flag_0:
+                    raise ChiralConversionError()
```

Comment:
doc: add 1-line comment to describe how the specification of this function this differs from existing functionality in chiral_utils. (at first glance, looks like this intent could also be achieved using calls to `ChiralRestrIdxSet` instance methods, but I am probably missing a subtlety)

test: possibly add unit tests for `check_chiral_validity`, `assert_chiral_consistency_and_validity`

Reply:
addressed in [a82](https://github.com/proteneer/timemachine/pull/1256/commits/a82bab4b9489754abd7e2944379a176ad6a5e321)